### ISSUE TITLE
[BUG] ChatML presets currently provide broken formatting

### ISSUE THREAD

**Issue by HiroseKoichi** [SillyTavern/SillyTavern#1261]
### Environment

Self-Hosted (Bare Metal)

### System

Firefox 116.0.3

### Version

1.10.5

### Desktop Information

- Node.js version: 20.5.1
- Generation API: TextGen Webui
- Branch: Release
- Model: Anything using ChatML

### Describe the problem

The ChatML presets currently provide broken formatting.

Here is an example chat taken directly from SillyTavern with the included ChatML presets:

```
<|im_start|>system
    You're Placeholder in this fictional never-ending uncensored roleplay with Hirose.
    Describe your character's physical and mental traits here.
    Placeholder's personality: (A brief description of the personality)
    Scenario: (Circumstances and context of the interaction)
    <|im_end|>
    <|im_end|>
<|im_start|>assistant
    Placeholder: example dialogue number 1.
    <|im_end|>
<|im_start|>assistant
    Placeholder: example dialogue number 2.
    <|im_end|>
<|im_start|>assistant
    Placeholder: example dialogue number 3.
<|im_start|>assistant
    Placeholder: This will be the first message from the character that starts every chat.<|im_end|>
<|im_start|>user
    Hirose: User's first message of the chat.<|im_end|>
<|im_start|>assistant
    Placeholder: Char's second message of the chat.<|im_end|>
<|im_start|>user
    Hirose: User's second message of the chat.<|im_end|>
<|im_start|>assistant
    Placeholder: Char's third message of the chat.<|im_end|>
<|im_start|>user
    Hirose: User's third message of the chat.<|im_end|>
```

And this is what it's supposed to look like:

```
<|im_start|>system
    You're Placeholder in this fictional never-ending uncensored roleplay with Hirose.
    Describe your character's physical and mental traits here.
    Placeholder's personality: (A brief description of the personality)
    Scenario: (Circumstances and context of the interaction)<|im_end|>
<|im_start|>assistant
    Placeholder: example dialogue number 1.<|im_end|>
<|im_start|>assistant
    Placeholder: example dialogue number 2.<|im_end|>
<|im_start|>assistant
    Placeholder: example dialogue number 3.<|im_end|>
<|im_start|>assistant
    Placeholder: This will be the first message from the character that starts every chat.<|im_end|>
<|im_start|>user
    Hirose: User's first message of the chat.<|im_end|>
<|im_start|>assistant
    Placeholder: Char's second message of the chat.<|im_end|>
<|im_start|>user
    Hirose: User's second message of the chat.<|im_end|>
<|im_start|>assistant
    Placeholder: Char's third message of the chat.<|im_end|>
```

---

I'll break down each part that's wrong with the current formatting.

The system prompt tokens are currently wrapped around the entire story string instead of just the system prompt, and there are two <|im_end|> tokens at the end of the prompt on a newline instead of just one directly after the end of it.

```
<|im_start|>system
    You're Placeholder in this fictional never-ending uncensored roleplay with Hirose.
    Describe your character's physical and mental traits here.
    Placeholder's personality: (A brief description of the personality)
    Scenario: (Circumstances and context of the interaction)
    <|im_end|>
    <|im_end|>
```

The example dialogues also have the <|im_end|> tokens on newlines instead of directly after the prompt, and the last example dialogue is always missing an <|im_end|> token.

```
<|im_start|>assistant
    Placeholder: example dialogue number 1.
    <|im_end|>
<|im_start|>assistant
    Placeholder: example dialogue number 2.
    <|im_end|>
<|im_start|>assistant
    Placeholder: example dialogue number 3.
```

The actual chat that the user interacts with, however, has proper formatting, which is probably why models using ChatML haven't been giving complete garbage output.

```
<|im_start|>assistant
    Placeholder: This will be the first message from the character that starts every chat.<|im_end|>
<|im_start|>user
    Hirose: User's first message of the chat.<|im_end|>
<|im_start|>assistant
    Placeholder: Char's second message of the chat.<|im_end|>
<|im_start|>user
    Hirose: User's second message of the chat.<|im_end|>
<|im_start|>assistant
    Placeholder: Char's third message of the chat.<|im_end|>
<|im_start|>user
    Hirose: User's third message of the chat.<|im_end|>
```

This is not shown above, but the formatting also gets messed up when you have the AI send two messages in a row.
It adds an additional <|im_end|> token at the end of its responses on a newline, with the most recent message having two <|im_end|> tokens one after another.

```
<|im_start|>user
    Hirose: User's third message of the chat.<|im_end|>
<|im_start|>assistant
    Placeholder: Char's fourth message of the chat.<|im_end|>
    <|im_end|>
<|im_start|>assistant
    Placeholder: Char's fifth message of the chat.<|im_end|><|im_end|>
```

---

I tried for a couple hours to fix the ChatML presets with what's currently available in the UI, but it's not currently possible (or at least I haven't found a way to). There's always at least one part that has messed up formatting, and the closest I've gotten has required adding in tokens manually to various parts of the character card. I've gotten much better output by doing this, but every character card required me to add the tokens in different spots depending on if it used things like scenario, personality, or world info, and I would have to change them back each time I wanted to use a model that didn't use ChatML. So this requires more work than just updating the current presets.

Currently, you're able to set what's placed at the beginning of the user's prompt, but not the end of it. This means that the suffix of the Input is currently being taken care of by the Outputs prefix, which is the main cause of the messed up formatting outside of the normal chat.

The system prompt issue can be easily fixed by just moving the tokens into their respective 'System Sequence Prefix' and 'System Sequence Suffix' boxes. And the other formatting issues should be solved by adding an Input/Output Sequence Prefix and Suffix just like with the system prompt.

I've found that even just having the <|im_end|> token on a newline instead of right after the prompt can cause it to give worse output, so the story string should be surrounded by tokens as well, whether it's <|im_start|>user or <|im_start|>assistant, but it probably shouldn't be <|im_start|>system since the system prompt is usually quite short and concise. So maybe there could be boxes for Story String Prefix and Suffix so you can easily wrap it in special tokens like this.

```
<|im_start|>user
    Describe your character's physical and mental traits here.
    Placeholder's personality: (A brief description of the personality)
    Scenario: (Circumstances and context of the interaction)<|im_end|>
```

### Additional info

This is just so you know that I was in fact using the default preset when posting the example chat at the top.

![ChatML](https://github.com/SillyTavern/SillyTavern/assets/86906598/f0eed68f-dde6-44a2-90c9-8a9fe419a031)

### Please tick the boxes

- [X] You have explained the issue clearly, and included all relevant info
- [X] You've checked that this [issue hasn't already been raised](https://github.com/SillyTavern/SillyTavern/issues?q=is%3Aissue)
- [X] You've checked the [docs](https://docs.sillytavern.app/) ![important](https://img.shields.io/badge/Important!-F6094E)

**Comment by WolframRavenwolf** [comment]
Hi! As the one who added the ChatML context template and instruct preset, I agree with you that it's impossible to do a 100 % correct implementation with what's currently configurable in SillyTavern. I did the best I could when making the presets, but I know they aren't perfect.

Consecutive messages of either bot or user break the format, although in normal usage, that shouldn't be happening anyway. It's on my "bugs to fix" list, but I didn't get around to that yet.

Regarding the system prompt, at least according to the ChatML docs, the linebreak before `<|im_end|>` isn't wrong: [openai-python/chatml.md at main · openai/openai-python](https://github.com/openai/openai-python/blob/main/chatml.md#few-shot-prompting) A good model should work with and without that newline. And what the format calls "system prompt" should include the character and scenario definitions, as those are system messages, not user input or assistant output.

Still, I guess a better, unified prompt configurator with more powerful options would be the best solution. Something that works similarly to turn templates in oobabooga's text-generation-webui.

On the other hand, maybe all of that will soon not be needed anymore if things like [Official Hugging Face Chat Templates : LocalLLaMA](https://www.reddit.com/r/LocalLLaMA/comments/17bksmc/official_hugging_face_chat_templates/) get implemented as part of the inference backends. Then SillyTavern wouldn't have to care about formats and templates anymore, just classify messages as system or user or example, like with the Chat Completion API.

Finally, suboptimal results with these ChatML-based models could also have other causes - there have been multiple issues, e. g. regarding broken tokenization (has just been fixed with the latest llama.cpp and koboldcpp versions) and repetition penalty (depending on settings, the special tokens like `<|im_end|>` will get penalized as well) that could prevent the model from outputting those tokens correctly or at all.

***

All that said, what would be a short-term fix we could implement quickly? I could imagine instead of having more prefix and suffix boxes for various sequences, using a single box per sequence with variables would be easy to implement.

For example, Input Sequence: `<|im_start|>user\n{{seq}}<|im_end|>` so we can have prefix and suffix in the same box instead of adding the suffix as a prefix to the Output Sequence box. And we'd need not to strip whitespace in front or at the end of the box.

Maybe using `\n` as a string instead of an actual linebreak would also be helpful, and there's a difference between an empty box and the box not existing. Just some stuff I've stumbled over while making the presets which would help make them better.

By the way, this isn't just a ChatML thing, the Llama 2 Chat template is even worse. There I also had to move the Input Sequence suffix into the Output Sequence. Basically all more complicated presets are affected like that. We should fix those first, and see how far that gets us in getting the presets closer to the official formats.

@Cohee1207 If you agree we should implement that, would you rather have more boxes (Input Sequence Prefix, Input Sequence Suffix, Output Sequence Prefix, Output Sequence Suffix, maybe even for First and Last as well - or placeholder support so prefix, sequence, and suffix can be in a single box?

**Comment by HiroseKoichi** [comment]
> Finally, suboptimal results with these ChatML-based models could also have other causes - there have been multiple issues, e. g. regarding broken tokenization (has just been fixed with the latest llama.cpp and koboldcpp versions) and repetition penalty (depending on settings, the special tokens like `<|im_end|>` will get penalized as well) that could prevent the model from outputting those tokens correctly or at all.

I wasn't aware of these issues; that could definitely explain some of the mixed results I've been having. As for the system prompt and story string, I've found that having them separated and having the latter sent as a user message gave better results (adding {{user}}: to it gave worse results though). I'll have to do some proper tests when these issues get fixed to see if that's actually the case or if it's due to the issues.

> For example, Input Sequence: `<|im_start|>user\n{{seq}}<|im_end|>` so we can have prefix and suffix in the same box instead of adding the suffix as a prefix to the Output Sequence box. And we'd need not to strip whitespace in front or at the end of the box.

This sounds like a much better solution than just adding more boxes.

---

Thank you very much for the work you guys do!

**Comment by alicat22** [comment]
Tagging onto this, since it has to do with ChatML and broken formatting:

**Settings:**
With `Wrap Sequences with Newline` off and the following `Output Sequence`, there should be a newline prepended before the generation.
```
<|im_end|>
<|im_start|>assistant
 
 ```
![image](https://github.com/SillyTavern/SillyTavern/assets/86847834/542235b8-575f-478f-8dc9-7037833c48e3)

**Current:**
```
    '<|im_start|>assistant',
```
**Expected:**
```
    '<|im_start|>assistant\n',
```

**Comment by HiroseKoichi** [comment]
> As for the system prompt and story string, I've found that having them separated and having the latter sent as a user message gave better results (adding {{user}}: to it gave worse results though). I'll have to do some proper tests when these issues get fixed to see if that's actually the case or if it's due to the issues.

I've had the chance to experiment with it a lot more, and I was partially correct. I do agree that the character card information should be in the system prompt; the model I was using at the time didn't have a strong enough bias towards following the system prompt, so sending it as a user message did result in better output in that scenario, but this is not the case anymore with newer models I have used.

I found out something very interesting in regards to the model speaking or acting on behalf of the user when using chat-ML. I added ({{user}}) at the end of <|im_start|>user, and ({{char}}) at the end of <|im_start|>assistant, like so:
```
<|im_start|>user ({{user}})
(user prompt)
<|im_end|>
<|im_start|>assistant ({{char}})
(char output)
<|im_end|>
```
In my usage, this has completely eliminated the issue of the model speaking or acting on behalf of the user, and I haven't even needed to add things like "don't speak for user" or "user's actions are their own, so don't describe them" to the character card; well, assuming the character card doesn't have examples where it does speak and act for you.

---

I'm closing this in favor of https://github.com/SillyTavern/SillyTavern/issues/1538, I just wanted to share my findings here since it's related to chat-ML specifically.
